molgenis-navigator/src/main/java/org/molgenis/navigator/copy/service/EntityTypeCopier.java

```diff
@@ -0,0 +1,245 @@
+import static java.util.stream.Collectors.toSet;
+import static java.util.stream.Stream.concat;
+import static org.molgenis.data.meta.model.EntityTypeMetadata.ENTITY_TYPE_META_DATA;
+import static org.molgenis.navigator.copy.service.RelationTransformer.transformExtends;
+import static org.molgenis.navigator.copy.service.RelationTransformer.transformMappedBys;
+import static org.molgenis.navigator.copy.service.RelationTransformer.transformPackage;
+import static org.molgenis.navigator.copy.service.RelationTransformer.transformRefEntities;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import org.molgenis.data.AbstractEntityDecorator;
+import org.molgenis.data.DataService;
+import org.molgenis.data.Entity;
+import org.molgenis.data.file.model.FileMeta;
+import org.molgenis.data.file.model.FileMetaMetaData;
+import org.molgenis.data.meta.EntityTypeDependencyResolver;
+import org.molgenis.data.meta.model.EntityType;
+import org.molgenis.data.meta.model.EntityTypeMetadata;
+import org.molgenis.data.populate.IdGenerator;
+import org.molgenis.data.util.EntityTypeUtils;
+import org.springframework.stereotype.Component;
+
+@Component
+class EntityTypeCopier {
+
+  private static final int BATCH_SIZE = 1000;
+
+  private final DataService dataService;
+  private final IdGenerator idGenerator;
+  private final EntityTypeDependencyResolver entityTypeDependencyResolver;
+  private final EntityTypeMetadataCopier entityTypeMetadataCopier;
+
+  EntityTypeCopier(
+      DataService dataService,
+      IdGenerator idGenerator,
+      EntityTypeDependencyResolver entityTypeDependencyResolver,
+      EntityTypeMetadataCopier entityTypeMetadataCopier) {
+    this.dataService = requireNonNull(dataService);
+    this.idGenerator = requireNonNull(idGenerator);
+    this.entityTypeDependencyResolver = requireNonNull(entityTypeDependencyResolver);
+    this.entityTypeMetadataCopier = requireNonNull(entityTypeMetadataCopier);
+  }
+
+  public void copy(List<EntityType> entityTypes, CopyState state) {
+    entityTypes.forEach(entityType -> assignUniqueLabel(entityType, state));
+    List<EntityType> copiedEntityTypes = copyEntityTypes(entityTypes, state);
+    persist(copiedEntityTypes, state);
+  }
+
+  private List<EntityType> copyEntityTypes(List<EntityType> entityTypes, CopyState state) {
+    List<EntityType> copiedEntityTypes =
+        concat(entityTypes.stream(), state.entityTypesInPackages().stream())
+            .map(original -> copyEntityType(original, state))
+            .collect(toList());
+
+    copiedEntityTypes.forEach(entityType -> transformRelations(entityType, state));
+    return copiedEntityTypes;
+  }
+
+  private void persist(List<EntityType> copiedEntityTypes, CopyState state) {
+    entityTypeDependencyResolver
+        .resolve(copiedEntityTypes)
+        .stream()
+        .map(copy -> cutDefaultValues(copy, state))
+        .map(this::persistEntityType)
+        .collect(toList())
```

Comment:
What is the purpose of collecting to a list while streaming? If this is required then please add a note to the source code.